site-builder.js

```diff
@@ -0,0 +1,156 @@
+// Author: Mike Bland (michael.bland@gsa.gov)
+/* jshint node: true */
+
+'use strict';
+
+var fs = require('fs');
+var path = require('path');
+var build_logger = require('./build-logger');
+var child_process = require('child_process');
+
+var exports = module.exports = {};
+
+// Executes the algorithm for cloning/syncing repos and publishing sites.
+// Patterned after the ControlFlow pattern used within Google.
+//
+// Once instantiated, users need only call build(), which is the entry point
+// to the algorithm. All other methods are "states" of the algorithm/state
+// machine that are executed asynchronously via callbacks.
+//
+// repo_dir: directory containing locally-cloned Pages repositories
+// repo_name: name of the repo belonging to the 18F GitHub organization
+// dest_dir: path to the destination directory for published sites
+// site_path: path to the repo on the local machine
+// branch: the branch to publish
+// build_logger: BuildLogger instance
+// done_callback: callback triggered when the algorithm exits; takes a single
+//   `err` argument which will be nil on success, and an error string on
+//   failure
+function SiteBuilder(repo_dir, repo_name, dest_dir, site_path, branch,
```

Comment:
I think the number of params here could be made clearer by using an options object. This would also allow there to be defaults:

```js
function SiteBuild(opts) {
  this.repo_dir = opts.repo_dir;
  this.repo_name = opts.repo_name

  // Or better but potentially dangerous
  Object.assign(this, opts); // Not sure what version of nodejs you're running but this is an es6 method.
}
```